### PR TITLE
add `isNull` condition for payload filtering

### PR DESCRIPTION
**Pull request by ibrahim-akrab**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

This PR will solves issue #1609 with some rough edges.

### TODO:
* [x] check the cardinality estimation correctness
* [x]  unit tests
* [x] OpenAPI integration tests
* [x] grpc implementation

/claim #1609

### CODE REVIEW

**Comment by ibrahim-akrab** [comment]
Hey @agourlay, sorry to bother but I've been trying to get grpc api to work for some time now. I edited `grpc/proto/qdrant.rs` according to the updates I made but for some reason it gets overwritten. Is there somewhere else that I need to update for it to be reflected in the grpc api?

**Comment by ibrahim-akrab** [comment]
Alright, now I think it should be ready (except for the cardinality estimation since I need some help to make sure it is right)

I found that `IsEmpty` condition only gets matched if the key is not present at all in the payload. It doesn't get matched when it is `[]` as it was mentioned in the original issue. I checked that against the docker image without my changes as well.  I think this is in part because of 
https://github.com/qdrant/qdrant/blob/ab03267b94b5042822c0a1aae1c3be95a35da628/lib/segment/src/common/utils.rs#L148-L151
Since it treats `serde_json::value::Value::Array` as a `Single` value. So this can be fixed but I am not sure that it doesn't have regressions elsewhere.
Another approach which I think will be less error prone is fixing https://github.com/qdrant/qdrant/blob/ab03267b94b5042822c0a1aae1c3be95a35da628/lib/segment/src/payload_storage/query_checker.rs#L94 to take the previous fact into consideration. 

I am willing to implement either fix of them. Just waiting for your opinion and approval.

**Comment by ibrahim-akrab** [comment]
I went ahead and added a fix for the `is_empty` condition of `"key":[]` mentioned above. Now it behaves according to the issue description. It can always be reverted if it wasn't meant to behave that way.

File: lib/segment/src/common/utils.rs
Comment (ibrahim-akrab) [inline lib/segment/src/common/utils.rs:1]: I added a specific implementation for the `MultiValue<&Value>` used in the `is_empty` and `is_null` conditions. I couldn't change the main `is_empty()` method since it's being used elsewhere and the new specialized methods couldn't have the same name since the compiler won't attempt to resolve the ambiguity in case they overlap.

**Comment by ibrahim-akrab** [comment]
Hey @agourlay, could you please review this and let me know of any required changes.

File: lib/segment/src/payload_storage/query_checker.rs
Comment (timvisee) [inline lib/segment/src/payload_storage/query_checker.rs:285]: Minor nitpick: you can probably just assert directly to reuse the same variable without a counter.

```rust
let is_null_condition = Filter::new_must(Condition::IsNull(IsNullCondition {
    is_null: PayloadField {
        key: "amount".to_string(),
    },
}));
assert!(!payload_checker.check(0, &is_null_condition));
```

**Review by timvisee — CHANGES_REQUESTED** [review]
Did a very basic review. Someone else will have to go over this in more detail.

File: lib/segment/src/payload_storage/query_checker.rs
Comment (ibrahim-akrab) [reply lib/segment/src/payload_storage/query_checker.rs:285]: And it will be more readable with each assertion after the assignment. Great idea, thanks.

**Review by timvisee — APPROVED** [review]

File: lib/segment/src/index/struct_payload_index.rs
Comment (ibrahim-akrab) [reply lib/segment/src/index/struct_payload_index.rs:269]: I didn't spend that much time around indices so I couldn't be sure about it. I will change it to match the `IsEmpty` estimations:
https://github.com/qdrant/qdrant/blob/6c4baae2624d067d4522dc110790c18bd36511d8/lib/segment/src/index/struct_payload_index.rs#L238-L240

File: lib/segment/src/payload_storage/query_checker.rs
Comment (agourlay) [inline lib/segment/src/payload_storage/query_checker.rs:244]: Is it possible to add a test for a field that is `[null]`?

File: lib/segment/src/types.rs
Comment (agourlay) [inline lib/segment/src/types.rs:1551]: this debug statement one can be removed

File: lib/segment/src/types.rs
Comment (agourlay) [inline lib/segment/src/types.rs:1578]: same

File: lib/segment/src/payload_storage/query_checker.rs
Comment (ibrahim-akrab) [reply lib/segment/src/payload_storage/query_checker.rs:244]: How is the filtering supposed to behave in that case?

Comment (agourlay) [reply lib/segment/src/payload_storage/query_checker.rs:244]: I guess `[null]` is equivalent to an empty array, so not null.
WDYT?

File: openapi/tests/openapi_integration/helpers/collection_setup.py
Comment (ibrahim-akrab) [reply openapi/tests/openapi_integration/helpers/collection_setup.py:52]: The `values_count` change is just an example. What I meant to say is without having the test data representing all cases, some change may break another one and never fail a test since they're not existing in the same collection.

I was already reverted the test changes to their previous state and added a new collection with the diverse payloads per @generall's request. However, I am not happy with this solution for the reasons mentioned above. I'd rather add to the old test points.

File: lib/segment/src/payload_storage/query_checker.rs
Comment (ibrahim-akrab) [reply lib/segment/src/payload_storage/query_checker.rs:244]: With the current implementation, it will not be matched with neither `is_empty` nor `is_null`.
If you want it to be matched with `is_empty` that's fine. However, what about `[null, [null]]`, is that even a valid payload? if so, should it be equivalent to empty as well?

Comment (agourlay) [reply lib/segment/src/payload_storage/query_checker.rs]: Forget about my comment, I don't think it is necessary to test this.
Sorry for derailing your research.

Comment (ibrahim-akrab) [reply lib/segment/src/payload_storage/query_checker.rs]: It's alright, I added it anyway to test for any future changes regarding this case.

**Review by generall — APPROVED** [review]
LGTM

**Comment by generall** [comment]
@ibrahim-akrab , thanks for the contribution! Please do not forget to `/claim #1609` the bounty before I merge this

**Comment by ibrahim-akrab** [comment]
@generall, happy to help and thanks for the bounty :money_mouth_face:

**Review by agourlay — APPROVED** [review]
Thanks for your contribution :+1: 

If you are still feeling motivated you could add a short doc about it in the [Filtering](https://qdrant.tech/documentation/filtering/#is-empty) section.

**Comment by ibrahim-akrab** [comment]
@agourlay Just submitted a PR there. Thanks for the suggestion :+1:

**Pull request merged**
